kernel/src/net/socket/ip/stream/mod.rs

```diff
@@ -651,11 +670,23 @@ impl Socket for StreamSocket {
     }
 
     fn set_option(&self, option: &dyn SocketOption) -> Result<()> {
-        let mut options = self.options.write();
+        let (mut options, mut state) = self.update_connecting();
 
-        match options.socket.set_option(option) {
+        match options.socket.set_option(option, state.as_mut()) {
             Err(err) if err.error() == Errno::ENOPROTOOPT => (),
-            res => return res,
+            Ok(need_iface_poll) => {
+                let iface_to_poll = need_iface_poll.then(|| state.iface().cloned()).flatten();
+
+                drop(state);
+                drop(options);
+
+                if let Some(iface) = iface_to_poll {
+                    iface.poll();
+                }
+
+                return Ok(());
+            }
+            res => return res.map(|_| ()),
```

Comment:
I think making the pattern explicit by writing `Err(err) => return Err(err)` should improve the readability of the code (otherwise the reader may have to think about whether `res` can be here).

Also, putting `Err(err) => return Err(err)` just below `Err(err) if err.error() == Errno::ENOPROTOOPT => (),` should be even better.